### PR TITLE
Duplication and atlas update

### PR DESCRIPTION
**Pull request by Hakazaba**
Duplication now will be taken account of during the generation of the atlas. This may need an update of Inochi session to be compatible and needs testing.

There are some variables that may not be currently copied over. Someone more familiar with the functions of composites for instance, should look over the variables to make sure everything important is copied.

### CODE REVIEW

File: source/creator/actions/node.d [hunk 1]
```diff
@@ -332,6 +332,76 @@ void incAddChildWithHistory(Node n, Node to, string name=null) {
     incActivePuppet().rescanNodes();
 }
 
+Node recursiveDuplicate(Node n){
+    Node x;
+
+    if ((cast(Part) n) is null) {
```
Comment (LunaTheFoxgirl) [inline source/creator/actions/node.d]: Please follow the code-style used throughout the rest of Inochi2D, this is difficult to read.

File: source/creator/actions/node.d [hunk 2]
```diff
@@ -332,6 +332,76 @@ void incAddChildWithHistory(Node n, Node to, string name=null) {
...
+    }
+    x.name = n.name;
+    x.enabled = n.enabled;
+    x.globalTransform = n.globalTransform;
+    x.localTransform = n.localTransform;
+    x.zSort = n.zSort;
+
+ 
+    foreach (child; n.children())
+    {
```
Comment (LunaTheFoxgirl) [inline source/creator/actions/node.d]: This curlybrace should be on the same line as the foreach.

File: source/creator/panels/nodes.d [hunk 3]
```diff
@@ -165,7 +167,25 @@ protected:
                 incText(incTypeIdToIcon("MeshGroup"));
                 igSameLine(0, 2);
                 if (igMenuItem(__("MeshGroup"), "", false, true)) incAddChildWithHistory(new MeshGroup(cast(Node)null), n);
-
+                
+                if((cast(Camera)n is null) & (cast(Puppet)n is null)){
```
Comment (LunaTheFoxgirl) [inline source/creator/panels/nodes.d]: This could be simplified to
```d
if (cast(Camera)n || cast(Puppet)n)
```

**Review by LunaTheFoxgirl — CHANGES_REQUESTED** [review]
Good work, however there's some cleanup to do before this can be merged.

File: source/creator/panels/nodes.d (hunk 3 above)
Comment (Hakazaba) [reply source/creator/panels/nodes.d]: We only want to proceed if they're both null. If they aren't then we might end up trying to copy the puppet or camera.

File: source/creator/actions/node.d (hunk 1 above)
Comment (Hakazaba) [reply source/creator/actions/node.d]: I'll try to clean it up a bit tomorrow

File: source/creator/panels/nodes.d (hunk 3 above)
Comment (LunaTheFoxgirl) [reply source/creator/panels/nodes.d]: In which case just do !cast; that will be a is-null check

**Review by LunaTheFoxgirl — APPROVED** [review]

**Pull request merged**